### PR TITLE
Added Dask Experiments

### LINKED ISSUE #1: Create new image classification experiments w/ GPUs
It'd be nice to test out the Dask implementation of batch size growth to see how the training time is affected.

I'm thinking that three damping techniques should be used:

1. Decreases the learning rate
2. Increases the batch size
3. Decreasing the learning rate initially, then increasing the batch size

@JoeHolt it'd be great if you could implement this experiment.

### PR DESCRIPTION
**Pull request by JoeHolt**
**What does this PR implement?**
It runs experiments with the WideResnet model on the CIFAR10 dataset. It mirrors the experiments in the GeoDamp paper.

**Reference issues/PRs**
This closes #1.

### CODE REVIEW

File: exp-dask/train.ipynb [hunk 1]
```diff
@@ -165,33 +166,101 @@
...
+      "Initial Learning Rate: 0.1\n",
+      "Epoch: 1 (acc: 0.09953999519348145)\n",
+      "Epoch: 2 (acc: 0.09959999471902847)\n",
+      "Epoch: 3 (acc: 0.09953999519348145)\n",
+      "Epoch: 4 (acc: 0.09963999688625336)\n",
+      "Epoch: 5 (acc: 0.09963999688625336)\n",
+      "Epoch: 6 (acc: 0.09978000074625015)\n",
+      "Epoch: 7 (acc: 0.09971999377012253)\n",
+      "Epoch: 8 (acc: 0.09933999925851822)\n",
+      "Epoch: 9 (acc: 0.09947999566793442)\n",
```
Comment (stsievert) [inline exp-dask/train.ipynb]: The accuracy is not increasing quickly enough. After 10 epochs, the accuracy should be at least 20%.

Is the loss changing equally slowly? It looks like it is. If so, I would increase the batch size until it's obvious that you shouldn't.

Is the loss/accuracy changing every model update? You should be able to see that data with this code:

``` python
# adadamp/_dist.py
def _update_model(model_opt, grads):
    model, optimizer = model_opt

    loss = sum(info["_loss"] for info in grads)
    print(loss)  # will print to dask worker logs (command line?)

    optimizer.zero_grad()
    for name, param in model.named_parameters():
        ...
```

I suspect the model and optimizer aren't hooked up properly. If that's the case (and only if that's the case), you might want to look at some of the code in [add_param_group][1].

[1]:https://github.com/pytorch/pytorch/blob/42d2e31cd6d798fe887559465452613378e4b821/torch/optim/optimizer.py#L54-L55

File: exp-dask/train.ipynb [hunk 2]
```diff
@@ -165,33 +166,101 @@
...
-    }
-   ],
+   "outputs": [],
    "source": [
-    "torch.cuda.is_available()"
+    "toCSV = model.curr_metas\n",
+    "with open('./exp-dask/exp1-decreaseingLR-const-workers-v0.csv', 'w', encoding='utf8', newline='') as output_file:\n",
+    "    fc = csv.DictWriter(output_file, fieldnames=toCSV[0].keys())\n",
+    "    fc.writeheader()\n",
+    "    fc.writerows(toCSV)"
```
Comment (stsievert) [inline exp-dask/train.ipynb]: Suggestion: try this code:

``` python
import pandas as pd
toCSV = ... # as defined, a list of dicts. [{"foo": 0, "bar": 1}, {"foo": -1, "bar": 3}, ...]
pd.DataFrame(toCSV).to_csv("exp-dask/...")
```

File: exp-dask/exp1-decreaseingLR-const-workers-v0.csv [hunk 3]
```diff
@@ -0,0 +1,3 @@
+n_updates,n_data,score__calls,partial_fit__calls,partial_fit__time,partial_fit__batch_size,score__acc,score__loss,score__time
+391,50048,1,1,57.44204807281494,128,0.10057999938726425,2.3577922216796874,24.38423752784729
+782,100096,2,2,57.691697120666504,128,0.10099999606609344,2.35768447265625,24.21728205680847
```
Comment (stsievert) [inline exp-dask/exp1-decreaseingLR-const-workers-v0.csv]: It looks like this is for CIFAR10. If so, this timing looks pretty good! 57 seconds per epoch (so 140ms per model update). That's certainly competitive with the benchmarks on the README.

**Comment by JoeHolt** [comment]
Loss goes up with batch size. Am I supposed to normalize loss over number of items in batch?

**Comment by stsievert** [comment]
Yes, the loss should be the mean/average loss of the examples in the batch.


    loss = mean(loss_fn(examples) for example in batch_examples)

Does that answer your question?

**Comment by JoeHolt** [comment]
I am now running the experiment again with averaged loss (implemented in adadamp repo PR). The graph for number of model updates vs training loss does not even seem close to that in the example (although, this is still using the data from the summed-loss model, not the averaged one). 

Edit: Also it seems like we are ending with ~30,000 model updates, but the graphs from the paper appear to have ~80,0000 updates. 

Edit 2: The figures mistakenly say 'train loss/accuracy'. They are showing data for the test set loss and accuracy

**Comment by JoeHolt** [comment]
Results with updated dist.py appear to be nearly identical to previous run.

**Comment by stsievert** [comment]
Hm... it appears to be unstable. Maybe try decreasing the learning rate by a factor of 5 to `lr=0.02`?

**Comment by JoeHolt** [comment]
Results are a lot smoother now, but accuracy remains similar to before. I feel like we have something wrong with our setup, it seems like in the paper they are around 80% test accuracy after just 5-10 epochs, and we never some close to that level of accuracy. 

Why is the accuracy taking such a large hit after the final batch size change?

**Comment by stsievert** [comment]
> it seems like in the paper they are around 80% test accuracy after just 5-10 epochs 

I think that should be the debug loop. A definite failure would be not reaching 75% accuracy after 5–10 epochs. That should enable much quicker iteration. I think you'll know when you've got the training configured correctly.

> I feel like we have something wrong with our setup .... Why is the accuracy taking such a large hit after the final batch size change?

Ah. I bet the dataset is ordered. See if this change fixes it:

``` diff
# _dist.py#_get_gradients, line 369 on master
-        idx = [i % len_dataset for i in range(start_idx, start_idx + batch_size)]
+        idx = self.random_state_.choice(len_dataset, size=batch_size, replace=False)
+        idx.sort()
```

I've also gone through your training notebook ([train.ipynb][2]). I'd make the following changes:

* I'd rerun this with a static batch size (128) and learning rate decay. If that implementation doesn't work, something's wrong.
* I'd use the dataset cleaning at [pytorch-cifar/main.py#L39][1], not the tutorial cleaning. The normalization weights will change, and also some of the horizontal flipping/etc.

Could you generated a [Dask performance report][3] when you rerun these experiments?

[3]:https://distributed.dask.org/en/latest/diagnosing-performance.html#performance-reports
[1]:https://github.com/kuangliu/pytorch-cifar/blob/7941bebc1207af338a98eb7cd32aa58582d8d003/main.py#L39
[2]:https://github.com/stsievert/adadamp-experiments/blob/d0f0039e87/exp-dask/viz.ipynb

**Comment by JoeHolt** [comment]
I need to update the visualizations to better match the paper, but the initial look at experiment data seems to show like it is working now. 

I will try to run the decreasing batch size experiment over night tonight.

**Comment by JoeHolt** [comment]
Finished running experiments! All are plotted in the viz notebook similar to how they were in the Dont Decay the Learning Rate paper.

**Comment by stsievert** [comment]
This is perfect. I think those CSV histories are perfect to start running simulations with.

**Comment by stsievert** [comment]
I ran some timings of https://github.com/stsievert/adadamp/pull/19 on a `g4dn.xlarge` EC2 machine, which has a NVIDIA T4 Tensor Core ([datasheet][1]). More detail on the performance is in https://github.com/stsievert/adadamp/pull/19#issuecomment-777881529, and more detail on the calculation is in `exp-dask/simulation-sleep-times`.

I think the sleeping time for the simulations should be with this code:

``` python
DEEPCOPY_TIME = 0.05855  # seconds
GRAD_TIME_128 = 0.07832  # seconds

def gradient(model_opt, ..., idx):
    sleep(DEEPCOPY_TIME)
    sleep(GRAD_TIME_128 * len(idx) / 128)
    return {k: _random_tensor(v.size) for k, v in model.named_parameters()}
```

I think two sets of simulations should be run: one with deepcopy and one without. That way we can say "this is the best this Dask wrapper can do." We might see that performance if when include https://github.com/saturncloud/dask-pytorch-ddp is (eventually) integrated.

[1]:file:///Users/scott/Downloads/t4-tensor-core-datasheet-951643.pdf

**Comment by JoeHolt** [comment]
Finally got simulation working and running without error. Right now I have it running with all sleeps set to 0.0000001s to make sure it can run all the way through without issue. If this works, I'll run again with the actual times. 

I looked over the timing notebook and am still a little confused about how you came across the timing numbers.

*Q1*
It looks like the timing per gradient was calculated for the decreasing-lr run, deriving the time from the CSV I generated. Is this correct?
- If so, does this mean that those timing numbers are specific to the decreasing-lr experiment (and therefor only applicable to the decreasing-lr simulation)?
 
*Q2*
I am also wondering why a special EC2 instance was needed to derive timing data from the CSV data. It seems to me like the calculations in Calculation.py could be computed pretty easily on a laptop alone. 

*Q3*
If we do in fact have a constant time per 128 gradients and time for deep copy across all the experiments, what is the benefit to having ran 4 separate experiments before, when just one would be enough to get the timing numbers?

**Comment by JoeHolt** [comment]
What graphs did you want for the simulation data again?

Here's what I have written but I think I am missing one:
- Time vs Accuracy
- Time vs Loss

**Comment by JoeHolt** [comment]
I am currently testing the simulator class (ie setting all sleeps to 0s for testing) through a full 200 epochs. I added a new visualization notebook with graphs for time VS accuracy and loss through the first 60 epochs.

Does everything look okay so far?

**Comment by stsievert** [comment]
Yeah, it looks good. The plots I'd like include `{time, n_updates, n_epochs} vs {loss, accuracy}`. I'm a little confused on two points:

1. Why is `partial_fit__time` approximately 40 secs/epoch in [dec-lr/results-ep60.csv][2]? Is some variable not set to 0?
2. In the visualization notebook [viz-sim.ipynb][1], why are the x-axis limits different between the plots – shouldn't the training time be the same?

This visualization code is simpler than the code you're using:

``` python
df["training_time"] = df["partial_fit__time"].cumsum()
ax = df.plot(x="training_time", y="score__acc")

ax.set_xlabel("Training time (s)")  # etc
```

If it runs successfully for a couple epochs, I say go ahead and run the simulation.

[1]:https://github.com/stsievert/adadamp-experiments/blob/6b951581d145c478c2a4ec2c453db4b81d2d5330/exp-dask/viz-sim.ipynb
[2]:https://github.com/stsievert/adadamp-experiments/blob/6b951581d145c478c2a4ec2c453db4b81d2d5330/exp-dask/sim-results/dec-lr/results-ep60.csv

**Comment by JoeHolt** [comment]
I am not sure whats going on with the partial_fit__time taking so long. It is accurate in that it is taking ~40s/epoch, but I do not know why it is taking so long. All SCORE_TIME, DEEPCOPY_TIME, GRAD_TIME_128 are all equal to 0 on that run. 

The time for each worker to get a gradient calculation with 0s sleep time is about 0.0022096633911132812s.

Should I try to run with dask performance report?

**Comment by JoeHolt** [comment]
Ran the performance report over 3 epochs of the simulator, should be in a file called sim-report.html.

- Simulating the gradient with 0 sleep time took 15.80s total and updating the model took 24.0s total over the 3 epochs. 
- Looks like there is around 20s of idle (?) time after the gradients are computed and the model is updated each time. How can I tell what is going on there?

Could this long time simply be due to me running it on a laptop? or do you think something else is going on?

**Comment by stsievert** [comment]
I'm interested in running these simulations with the current software and the improved version of the software (https://github.com/stsievert/adadamp/issues/21). This will require running two sets of simulations:

* One with `DEEPCOPY_TIME = 58.55e-3` (the current value)
* One with `DEEPCOPY_TIME = 1.55e-3 * 2 * np.log2(P)` when there are `P` machines.

<details>
<summary>Details on the calculation</summary>

* Latency: 0.05ms (logged into WID optimization cluster, pinged adjacent machine in 0.1ms (time to send+receive one byte).
* Bandwidth: 54.54Gb/s, the bandwidth of a 2011 infiniband interconnect with 4 channels, which is common "on most systems" (sources: [wiki][inf], [archived roadmap graph][roadmap]).
* Model size: 10.49MB.

[inf]:https://en.wikipedia.org/wiki/InfiniBand
[roadmap]:https://web.archive.org/web/20110929111021/http://www.infinibandta.org/content/pages.php?pg=technology_overview

```
[ins] In [14]: m = model.Wide_ResNet(16, 4, 0.3, 10)
[ins] In [15]: nparams = sum([p.nelement() for p in m.parameters()])
[ins] In [16]: mb = nparams * 4 / 1024**2  # megabytes
[ins] In [17]: mb
Out[17]: 10.494789123535156
```

That means that one communication will take 0.819 + 0.05ms = 0.869ms:

``` python
>>> bandwidth = 54.54 * 1024 / 8 # megabytes / s
>>> t_one_comm = mb / bandwidth
>>> t_one_comm / 1e-3
1.5033102315294906
```

The number of communications for a tree all-reduce communication scheme is `log2(P)` with `P` machines/Dask workers. So, `DEEPCOPY_TIME = 1.55e-3 * 2 * np.log2(P)`.

</details>

**Comment by JoeHolt** [comment]
Looks like it is taking ~22s an epoch on GPU, which is ~2-3x increase from laptop, but not as quick as we were hoping.

Update: i forgot to push my local changes where I added the preprocessing. With this in place it is only taking ~7s/epoch with timings set to 0.

**Comment by JoeHolt** [comment]
Just ran 100 epochs with the new experiment. This experiment did not have the parameter based max batch size implemented (as discussed in the issue), but it did not get to the point where the batch size was large enough for this to matter (ie 120th epoch).

After 88th epoch, the accuracy dropped from consistently being around 0.8925999999046326 to a much lower value around 0.1-0.2. What could be causing this?

**Comment by stsievert** [comment]
> After 88th epoch, the accuracy dropped from consistently being around 0.8925999999046326 to a much lower value around 0.1-0.2. What could be causing this?
Merge state

Hm... that's strange. Maybe try `lr *= 2` and `bs *= 2` instead of `*= 5`? Maybe try a different `random_state`?

---

I've made some updates to adadamp. Turns out `max_batch_size` wasn't scaling the learning rat5e, and the worker max batch size was hard-coded.

I've added some code to set the "damping" level. This will automatically scale the learning rate/batch size and respect `max_batch_size`.

``` python
est = DaskClassifierExperiments(..., lr=orig_lr * 2, batch_size=256, max_batch_size=5120)

dampings = {
    0: 256,
    60: 256 * 5,
    120: 256 * 5 * 5,
    180: 256 * 5 * 5 * 5,
}

for epoch in range(60):
    bs = dampings.get(epoch, None)
    if bs:
        est.damping_ = bs
```

If `max_batch_size` is not set, the `damping_` is always equal to the batch size (which is why `bs` is used above). If `max_batch_size == batch_size`, the learning rate only decays.

**Comment by JoeHolt** [comment]
Sounds good. I pulled your changes and updated the training accordingly. I am running the experiment again now.

**Comment by JoeHolt** [comment]
Added new experiments and visualizations (On legend, new experiments are titled Big BS 1 and Big BS 2, big BS 2 is the one with 0.98 momentum). Looks like significantly less model updates were required.

**Comment by JoeHolt** [comment]
I think (knock on wood) the simulator is all set up and working. I am going to run it non stop until we meet tomorrow, hopefully 5 or 6 of the simulations will be done by then.

For the last 6 experiments, with `DEEPCOPY_TIME = 5.17e-3 * np.log2(P)`, what value of P should I use?

Sidenote:
Are there any good ways to keep my jupyter lab connection open and active, even if I close my laptop? Right now if my laptop closes while connected to an EC2 running jupyter, I loose connection.

**Comment by stsievert** [comment]
That's great to hear! I'm looking forward to seeing the results tomorrow.

I've been editing https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-785392333, and now `DEEPCOPY_TIME = 1.55e-3 * 2 * np.log2(P)`. The value of `P` should be the number of Dask workers, so `DEEPCOPY_TIME` is not a constant anymore.

> Are there any good ways to keep my jupyter lab connection open and active

Not really. I'll mention some workarounds tomorrow.

I've had some difficulty in the past with my computer falling asleep and canceling network connections. On macOS, setting "turn display off after [infinite time]" in System Preferences > Energy Adapter resolves this. I leave my laptop open all night, and turn the display brightness to the lowest setting.

As we talked about, the method I recommended only editing files on your own machine, and using the cluster only for executing those files. That means I use this debug loop:

1. On **local** machine: edit files.
2. On **local** machine: run `scp foo.py user@machine:~`
3. On **remote** machine: run `python foo.py` (which writes results to `bar/`).
4. On **local** machine: run `scp -r user@machine:~/bar/`
5. On **local** machine: look at results and go back to step 1.

I found that this method prevents accidentally removing/overwriting files. In practice, this method certainly saves hours, maybe days.

**Comment by JoeHolt** [comment]
With the new changes, 2 of the simulations have completed so far, decreasing LR and the large BS exp (In the middle of the night it threw an unexpected `CancelledError` and stopped running. I am not sure why, but I rebooted the EC2 and am running the rest now).

I ploted the results in `sim-viz.ipynb` and the results look good.

**Comment by JoeHolt** [comment]
I ran and visualized more of the simulations. It does not seem like the learning rate is being changed for the experiments where we do not reach the max batch size of ~5000. Am I supposed to be setting max BS differently for each sim?

File: exp-dask/classifier.py [hunk 4]
```diff
@@ -116,9 +116,11 @@ def _get_gradients(
         # Distribute the computation of the gradient. In practice this will
         # mean (say) 4 GPUs to accelerate the gradient computation. Right now
         # for ease it's a small network that doesn't need much acceleration.
-        deep_time = DEEPCOPY_TIME
+        deep_time = DEEPCOPY_TIME # 0.05855
         if MUTIPLE_MACHINES:
-            deep_time = deep_time * 2 * np.log2(self._sim_data['n_workers'])
+            # DEEPCOPY_TIME = 1.55e-3 * np.log2(P)
+            deep_time = 1.55 * (10**-3) * np.log2(self._sim_data['n_workers'])
```
Comment (stsievert) [inline exp-dask/classifier.py]:
```suggestion
            deep_time = 1.55e-3 * np.log2(self._sim_data['n_workers'])
```

Comment (JoeHolt) [reply exp-dask/classifier.py]: 👍

**Comment by JoeHolt** [comment]
The update to deep copy time appears to be working; large batch size simulation finished in 2725.880440711975 seconds.

**Comment by JoeHolt** [comment]
The decreasing LR experiment just finished, the training time was down from 12,000 seconds (with the constant ~0.05 deep copy time from the first set of experiments) to 7,000. Is this expected?

Edit: this makes sense to me, because the deep copy time for this scenario (1 worked) would be `1.55*1*10^-3 = .00155` which is significantly faster than the constant time of 0.05855.

**Comment by stsievert** [comment]
> 12,000 seconds (with the constant ~0.05 deep copy time from the first set of experiments) to 7,000. Is this expected?

Yes. That's the time required to do 200 epochs worth of gradient calculation. The fact that our software takes 12k seconds instead of 7k seconds for one worker is a sign that the software needs improvement. This software should not add any additional overhead, but apparently it adds a rather significant amount.

Also, `log2(1) == 0`, so the deepcopy time for 1 worker is 0 seconds – at it should be.

**Comment by JoeHolt** [comment]
Uploaded the new sim data and their new plots in latest commit

**Comment by stsievert** [comment]
It'd be great to run two sets of experiments:


1. Another training run with a constant batch size 512 and the same step size decay schedule (`damping = [5, 5**2, 5**3]` at epochs [60, 120, 180]).
2. One that has a faster network bandwidth than (simulated) in https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-785392333.

With (1), then we can simulate a different number of workers (`grad_per_worker in [32, 64, 128, 256, 512]`) to see how that affects training time.

(2) can be enabled with

``` python
DEEPCOPY_TIME = 0.127488e-3 * np.log2(P)  # seconds
GRAD_TIME_128 = 78.32e-3  # seconds
```

`GRAD_TIME_128` is from https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-777882730.

<details>
<summary>Details on <code>DEEPCOPY_TIME</code></summary>

* Bandwidth: 100GB/s (from Sec. 5 of [1]).
* Communication overlap for half computation time: Fig. 6 of [1].
* Latency: 0.025ms. (from from https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-785392333)
* Model size: 10.4947MB (from https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-785392333)
* Models sent on each communication: 2.

``` python
>>> bandwidth = 100 * 1024  # MB/s
>>> latency = 0.025e-3  # seconds
>>> model_size = 10.494789123535156  # MB
>>> models_sent = 2 * 0.5  # 2 models sent (w[k] and w[k-1]) and overlaps 50% of the time.
>>> assert models_sent == 1
>>> comm_time = model_size / bandwidth * models_sent
>>> sleep_time = (comm_time + latency)
>>> sleep_time / 1e-3
0.127488175034523
```

References:

1. "PyTorch Distributed: Experiences on Accelerating Data Parallel Training." Li et al. (including PyTorch core developers). https://arxiv.org/pdf/2006.15704.pdf

</details>

**Comment by JoeHolt** [comment]
I am running the (1) experiment you mentioned now. I am confused however on how the dampening works since the changes you made to _dist.py. Before I would set the batch size and LR directly on the given epochs with set_lr and set_bs, these methods were removed.

Now with the the `model.damping_ = bs` I am confused about how it would know to decrease the Lr at a set rate every 60 epochs. The batch size will always be 512, and I am setting the max bs to 512, but how do I specify how often to update the LR and what to set it to?

Edit: After reading the code a little closer, it seems like the LR change factor is being derived from how relatively large of a BS increase is occurring, is this correct?

**Comment by JoeHolt** [comment]
Follow up Question: Do you think it is okay if I run the (1) experiment on the EC2 GPU instance at the same time I run the simulations from (2)? Or will that mess with the sim timings?

**Comment by stsievert** [comment]
> the LR change factor is being derived from how relatively large of a BS increase is occurring

Yes. This will decrease the learning rate by a factor of 2:

``` python
static_bs = 128
est = DaskBaseClassifier(..., max_batch_size=static_bs, batch_size=static_bs, lr=1)
est.damping_ = static_bs * 2
print(est.lr_)  # 0.5
```

(I might have some details of this code wrong, but the spirit is definitely right).

> Do you think it is okay if I run the (1) experiment on the EC2 GPU instance at the same time I run the simulations from (2)? Or will that mess with the sim timings?

It'll definitely affect the GPU memory. I'm not sure the timing will be too affected (and we can always check after the fact). I'd run them serially if possible.

**Comment by JoeHolt** [comment]
> It'd be great to run two sets of experiments:
> 
>     1. Another training run with a constant batch size 512 and the same step size decay schedule (`damping = [5, 5**2, 5**3]` at epochs [60, 120, 180]).
> 
>     2. One that has a faster network bandwidth than (simulated) in [#2 (comment)](https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-785392333).
> 
> 
> With (1), then we can simulate a different number of workers (`grad_per_worker in [32, 64, 128, 256, 512]`) to see how that affects training time.
> 
> (2) can be enabled with
> 
> ```python
> DEEPCOPY_TIME = 0.127488e-3 * np.log2(P)  # seconds
> GRAD_TIME_128 = 78.32e-3  # seconds
> ```
> 
> `GRAD_TIME_128` is from [#2 (comment)](https://github.com/stsievert/adadamp-experiments/pull/2#issuecomment-777882730).
> Details on `DEEPCOPY_TIME`

What should the timings be for the simulations for (1). Should it be with multiple workers?

Update:
Currently running them with these timings:

    - DEEPCOPY_TIME = 1.55e-3 * 2 * np.log2(P) when there are P machines.
    - GRAD_TIME_128 = 78.32e-3

**Comment by JoeHolt** [comment]
Cleaned up this PR. I ended up redo-ing the file structure for the training and simulation CSVs. I tried to do it in a way that would make it easy for you to continue using your visualization code without too big of changes.

**Comment by stsievert** [comment]
Also for the re-run of scaling, could you set `GRAD_TIME_128` to reflect a better GPU? I think dividing by two would be sufficient:

``` python
GRAD_TIME_128 = 0.5 * 78.32e-3
DEEPCOPY_TIME = 0.127488e-3 * np.log2(P)  # seconds
```

This would simulate using a NVIDIA P100 GPU instead of a NVIDIA T4 GPU. P100s are higher powered training GPUs, T4s are low powered inference GPUs.

<details>
<summary>Why this is a approximation reasonable?</summary>
I believe this is at least a decent approximation because of these stats:

|   | T4 | P100 | Ratio |
|----:|:-----:|:---:|:---:|
| Single-precision TFLOPS | 8.1 | 10.6 | 0.76 |
| Memory bandwidth | 300 GB/s | 732 GB/s | 0.41 |
| Power | 70W | 300W | 0.23|

Pulled from the datasheets of [NVIDIA T4][T4] and [NVIDIA P100][P100].

</details>


[P100]:https://www.nvidia.com/en-us/data-center/tesla-p100/
[T4]:https://www.nvidia.com/en-us/data-center/tesla-t4/

**Comment by JoeHolt** [comment]
I am having trouble configuring the client as you described. When ever I set up the client this way, it spams this error `distributed.worker - WARNING - Worker is at 97% memory usage. Pausing worker.  Process memory: 1.00 GB -- Worker memory limit: 1.03 GB` to the console until the page crashes. The current sim.ipynb has an example of this. 

How can I fix / work around this?

**Comment by stsievert** [comment]
Hm... when did the errors appear? Just after initialization, before any training? Do the warnings still happen if you comment out `from model import Wide_ResNet`/`client.upload_file("./exp-dask/model.py")`?

I think another way to phrase this question is "what's eating up 1GB of memory?" Try removing the autoreload cell at the top.

**Comment by JoeHolt** [comment]
I have done some more debugging and the problem appears to be with the import statements. If I comment out the model upload and auto reload, the same error occurs. 

If I run this code before running any other cells, the error does not occur:
```python3
# import
from dask.distributed import Client

client = Client(processes=False, n_workers=16, threads_per_worker=1)

client
```
But then when I run the import cell (seen bellow) next, after creating the client, the spamming error occurs:
```python3
import os
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import csv
import importlib
import pandas as pd
from time import time
from copy import copy
```

**Comment by JoeHolt** [comment]
If I run this code alone (ie with the model upload, but none of the other import statements):
```python3
# import
from dask.distributed import Client
# training client
def _prep():
    from distributed.protocol import torch

client = Client(processes=False, n_workers=16, threads_per_worker=1)
client.run(_prep)

#from model import Wide_ResNet
client.upload_file("./exp-dask/model.py")

client
```
It spams this error: `distributed.worker - WARNING - Worker is at 93% memory usage. Pausing worker.  Process memory: 966.74 MB -- Worker memory limit: 1.03 GB`, similar to before with slightly lower memory usage

**Comment by stsievert** [comment]
Looks like this is a temporary error. Let's try running with `g4dn.2xlarge` instance, which has twice the memory (2GB/core, which is typical for numerical work).

**Comment by JoeHolt** [comment]
I attempted to run those last sims on the larger machine, but the same issue appeared. It seems like the memory usage per worker is going up over time (seen from the current `sim.ipynb` file's logs), I am not sure what would be causing this.

**Comment by stsievert** [comment]
I've run `sim.ipynb` on a `g4dn.2xlarge` instance, and got it working. Here's what I did:

1. Removed the cell with `%load_ext autoreload`
2. Ran `htop` and `nvidia-smi` to make sure memory usage didn't explode.
3. It didn't explode for me. I'd suggest you run the same commands (running `htop` and `watch nvidia-smi` in two tmux panes will do it).
3. Used `Client(processes=False, n_workers=16, threads_per_worker=1, memory_limit="32GB")`. This silences the warnings.

I'm only okay silencing the warnings because this is a simulation and the memory never explodes. I would monitor the memory throughout the experiment to make sure it doesn't increase. Luckily, the dataset is never used (the memory address is passed around, but it's never shuffled or queried).

**Comment by JoeHolt** [comment]
Re-running the simulations now based on your comment. Should be done by morning if no error comes up.

**Pull request merged**